Added output file pattern matching

Lethja opened the pull request:
Adds and documents printf like parameters to the `--outputfile-pattern` argument. This allows identical changes multiple anim files to be done in a very trivial way

Example:
```bash
mkdir left right
mv *.anim left/
python animDump  --mirror -o %p/../right/%n left/*.anim
```

Lethja committed:
Added output file pattern matching

tapple commented:
Neat. I had an undocumented template replacement of {} with the stem. printf style is probably more usual, yeah. 

Only suggested change is that I think lowercase format codes are more common: `%p` vs `%P`. I just checked what datetime does with their format codes, and they are lowercase

tapple commented on animDump.py:
```diff
@@ -710,7 +718,11 @@ def output_filename(input_filename):
+                                                           '%%N' = input file name,
+                                                           '%%P' = input file directory,
+                                                           '%%%%' = literal '%%'.
+                                                           File extensions will be appended automatically""")
```
The help message doesn't include the line breaks:
```
  --outputfile-pattern OUTPUTFILE_PATTERN, -o OUTPUTFILE_PATTERN
                        Output name with optional pattern matching: '%N' = input file name, '%P' = input file
                        directory, '%%' = literal '%'. File extensions will be appended automatically
```

you can add `formatter_class=argparse.RawDescriptionHelpFormatter,` if you want those: https://docs.python.org/3/library/argparse.html#formatter-class

Lethja replied:
`RawDescriptionHelpFormatter` doesn't multi-line `help` so I moved the pattern documentation to `epilog` instead

```
usage: animDump.py [-h] [--verbose] [--no-sort] [--markdown] [--outputfile-pattern OUTPUTFILE_PATTERN] [--time-scale ACTIONS]
                   [--frame-rate ACTIONS] [--offset ACTIONS [ACTIONS ...]] [--mirror] [--scale ACTIONS] [--joint-pri ACTIONS ACTIONS]
                   [--pri ACTIONS] [--ease-in ACTIONS] [--ease-out ACTIONS] [--loop ACTIONS] [--loop-start ACTIONS] [--loop-end ACTIONS]
                   [--drop-loc [ACTIONS ...]] [--drop-rot [ACTIONS ...]] [--drop-pri [ACTIONS ...]] [--drop-joint [ACTIONS ...]]
                   [--drop-empty-joints] [--add-constraint ACTIONS ACTIONS ACTIONS] [--drop-constraints] [--c-plane]
                   [--c-ease ACTIONS ACTIONS ACTIONS ACTIONS] [--c-source-offset ACTIONS ACTIONS ACTIONS]
                   [--c-target-offset ACTIONS ACTIONS ACTIONS] [--c-target-dir ACTIONS ACTIONS ACTIONS]
                   files [files ...]

Manipulate Secondlife .anim files

positional arguments:
  files                 anim files to dump or process

options:
  -h, --help            show this help message and exit
  --verbose, -v
  --no-sort, -U         when printing joints with -v, show in file order rather than abc order
  --markdown, --md      output in markdown table
  --outputfile-pattern OUTPUTFILE_PATTERN, -o OUTPUTFILE_PATTERN
                        Output anim file path/name, see 'output file pattern flags' below for details
  --time-scale ACTIONS, --tscale ACTIONS, --speed ACTIONS, -s ACTIONS
                        Adjust duration by the given factor eg 2.0 for half-speed/double duration, or 0.5 for double speed/half duration
  --frame-rate ACTIONS, --fps ACTIONS
  --offset ACTIONS [ACTIONS ...], --adjust ACTIONS [ACTIONS ...]
                        Adjust joint location [joint] [x y] z. joint is mPelvis if omitted. x, y are 0 if omitted
  --mirror, --flip
  --scale ACTIONS       Scale location keys; eg 2.0 for double-size avatar, 0.5 for half-size avatar
  --joint-pri ACTIONS ACTIONS
  --pri ACTIONS
  --ease-in ACTIONS
  --ease-out ACTIONS
  --loop ACTIONS
  --loop-start ACTIONS
  --loop-end ACTIONS
  --drop-loc [ACTIONS ...]
  --drop-rot [ACTIONS ...]
  --drop-pri [ACTIONS ...]
  --drop-joint [ACTIONS ...]
  --drop-empty-joints
  --add-constraint ACTIONS ACTIONS ACTIONS
  --drop-constraints
  --c-plane
  --c-ease ACTIONS ACTIONS ACTIONS ACTIONS
  --c-source-offset ACTIONS ACTIONS ACTIONS
  --c-target-offset ACTIONS ACTIONS ACTIONS
  --c-target-dir ACTIONS ACTIONS ACTIONS

output file pattern flags:
  %n: input file name
  %p: input file directory
  %%: literal '%'
  file extensions will be appended automatically
```

tapple commented on animDump.py:
```diff
@@ -696,8 +696,16 @@ def __call__(self, parser, namespace, values, option_string=None):
+                       .replace("%P", str(input_path.parent.resolve()))
+                       .replace("%%", "%"),
+                       parents=input_path.parents,
+                       suffix=input_path.suffix)
```
Neat. I had an undocumented template replacement of `{}` with the stem. printf style is probably more usual, yeah. 

Only suggested change is that I think lowercase format codes are more common: `%p` vs `%P`. I just checked what datetime does with their format codes, and they are lowercase

Lethja replied:
My rationale for using capitals was that a typing user would already be holding down shift to type a `%` character, but you're right, printf flags tend to be lowercase. Changed in e67500589a25b353998e267df008a7f28d509cfd

tapple commented on animDump.py:
```diff
@@ -696,8 +696,16 @@ def __call__(self, parser, namespace, values, option_string=None):
+                       .replace("%P", str(input_path.parent.resolve()))
+                       .replace("%%", "%"),
+                       parents=input_path.parents,
+                       suffix=input_path.suffix)
```
When I try it, the `suffix=` keyword argument to Path doesn't do anything:
```
In [14]: Path("this", suffix="anim")
Out[14]: PosixPath('this')
```

Here's how to do it:
```
In [15]: Path("this").with_suffix(".anim")
Out[15]: PosixPath('this.anim')
```

tapple replied:
looks like Path ignoring kwargs is a bug and it raises a DeprecationWarning starting in python 3.12: https://github.com/python/cpython/issues/74033

```
Python 3.12.1 (main, Feb  9 2024, 07:46:47) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pathlib import Path
>>> Path("this", suffix=None)
<stdin>:1: DeprecationWarning: support for supplying keyword arguments to pathlib.PurePath is deprecated and scheduled for removal in Python 3.14
PosixPath('this')
```

Lethja replied:
Updated to use `.with_suffix(input_path.suffix)` in e675005

tapple commented on animDump.py:
```diff
@@ -696,8 +696,16 @@ def __call__(self, parser, namespace, values, option_string=None):
+                       .replace("%N", input_path.stem)
+                       .replace("%P", str(input_path.parent.resolve()))
+                       .replace("%%", "%"),
+                       parents=input_path.parents,
```
When I try it, the `parents=` keyword argument to Path doesn't do anything:
```
In [16]: p = Path("this").resolve()

In [18]: Path("this", parents=p.parents)
Out[18]: PosixPath('this')
```

I'm not actually sure what you're trying to do with this line. I'd just remove it

Lethja committed:
Lowercase the output file name tokens

Moved output file name token documentation to epilog to make it multi-line

Clean up output_path initialization

tapple commented:
merged. I messed around with help text formatting and figured out that RawHelpTextFormatter is the answer: https://github.com/tapple/tanimbomb/commit/41e12e08b709d43afccab35d9cbb3d997c891c89
```
  --outputfile-pattern OUTPUTFILE_PATTERN, -o OUTPUTFILE_PATTERN
                        Output anim file path/name, with template substitution:
                            %n: input file name
                            %p: input file directory
                            %%: literal '%'
                        File extension will be appended automatically
  --time-scale ACTIONS, --tscale ACTIONS, --speed ACTIONS, -s ACTIONS
                        Adjust duration by the given factor eg:
                            2.0 for half-speed/double duration, or
                            0.5 for double speed/half duration
```

Lethja commented:
> merged. I messed around with help text formatting and figured out that RawHelpTextFormatter is the answer: [41e12e0](https://github.com/tapple/tanimbomb/commit/41e12e08b709d43afccab35d9cbb3d997c891c89)
> 
> ```
>   --outputfile-pattern OUTPUTFILE_PATTERN, -o OUTPUTFILE_PATTERN
>                         Output anim file path/name, with template substitution:
>                             %n: input file name
>                             %p: input file directory
>                             %%: literal '%'
>                         File extension will be appended automatically
>   --time-scale ACTIONS, --tscale ACTIONS, --speed ACTIONS, -s ACTIONS
>                         Adjust duration by the given factor eg:
>                             2.0 for half-speed/double duration, or
>                             0.5 for double speed/half duration
> ```

Great! I saw `RawHelpTextFormatter` in the documentation myself but the "raw" in the name made me think it would require indenting every help string manually. I'm glad that's not the case
